src/js/load-package.ts

```diff
@@ -160,7 +171,18 @@ async function downloadPackage(
   } else {
     file_name = channel;
   }
-  return await _loadBinaryFile(baseURL, file_name);
+  try {
+    return await _loadBinaryFile(baseURL, file_name);
+  } catch (e) {
+    if (!IN_NODE) {
+      throw e;
+    }
+  }
+  // If we are IN_NODE, download the package from the cdn, then stash it into
+  // the node_modules directory for future use.
+  let binary = await _loadBinaryFile(cdnURL, file_name);
+  await nodeFsPromisesMod.writeFile(`${baseURL}${file_name}`, binary);
```

Comment:
I'm not very sure of this. IMO this kind of behavior should better be more explicit. For example, `puppeteer` downloads binaries during installation (though that is not what we want because Pyodide is not only for server-side usage), or `playwright` provides a command `playwright install` to download binaries after installation.

Reply:
Maybe take a config variable? I guess my main concern is that `packages.json` is still going to include all of these packages but the request will fail when it tries to open the file which doesn't exist.

Reply:
> My main concern is that `packages.json` is still going to include all of these packages but the request will fail when it tries to open the file which doesn't exist.

Yes, I also think it would be great if people don't need to download packages manually. So providing a way of automatically downloading packages would be nice for the user experience.

My opinion is that we should let Pyodide users know what is going on. They need to know that `whl` files are not included in the npm packages and those will be downloaded after installation.

If we are going to download packages in runtime as in this PR, I think there needs a proper log telling that it is downloading and stashing packages.

> Maybe take a config variable

Yeah, a config variable for deciding whether to download missing packages in runtime (or die with error) would be nice IMO. It would also be nice if CDN URLs can be configured at the top level as there [were some requests](https://github.com/pyodide/pyodide/issues/2250).
Still, I think it would be great to provide an explicit way to install packages before runtime (maybe `$ npx pyodide install` or something?) and recommend that way. The advantages of downloading them before runtime will be:

- If there is a problem in the CDN, it can be checked beforehand.
- Running time is stabilized (we don't need to download a large package at the first run)
- We can verify checksums of packages beforehand.

Reply:
I am not likely to work on this myself, but I am happy if someone else would do it. In the meantime, I'm not sure what we should do in this PR. If you have an opinion, I would be happy for you to put in code that you like (you can push it directly to this branch). I think I agree with you about the long term solution, but I'm not what the best intermediate solutions are.

> I think there needs a proper log telling that it is downloading and stashing packages.

Perhaps adding logging to this for now would be okay? Or we could remove it from this PR and throw an error if a missing package is loaded?

Reply:
> Perhaps adding logging to this for now would be okay?

Yeah I am okay with that for this PR. Perhaps let's not document this behavior for now. Thanks for working on this!

> I am not likely to work on this myself too soon, but I am happy if someone else would do it.

I'll open a issue so we can revisit this later.